src/measure.rs

```diff
@@ -0,0 +1,174 @@
+/// * `font_bytes`: The font file bytes.
+///
+/// # Returns
+///
+/// The height of the text in pixels.
+pub fn measure_height_px(
+    text: String,
+    font_size: f32,
+    line_spacing: f32,
+    width: f32,
+    font_bytes: &[u8],
+) -> Result<f32, Box<dyn std::error::Error>> {
+    let harf_font_ref = FontRef::new(&font_bytes).expect("For font files to be font files!");
+    let skrifa_font_ref = SkrifaFontRef::new(&font_bytes).expect("Fonts to be fonts");
+
+    let metrics = skrifa_font_ref.metrics(Size::new(font_size), LocationRef::default());
+    let line_height = (metrics.ascent - metrics.descent + metrics.leading) * line_spacing;
+
+    let mut all_lines = Vec::new();
+    for text_line in text.lines() {
+        let mut lines = Vec::new();
+        let mut current_line = String::new();
+
+        // TODO: splitting whitespace may not be right for \t or other special characters.
+        for word in text_line.split_whitespace() {
+            let potential_line = if current_line.is_empty() {
+                word.to_string()
+            } else {
+                format!("{} {}", current_line, word)
+            };
+
+            if get_text_width(&potential_line, &harf_font_ref, &skrifa_font_ref, font_size) <= width
+            {
+                current_line = potential_line;
+            } else {
+                let should_break_word = current_line.is_empty() || potential_line.contains(" ");
+
+                if !current_line.is_empty() {
+                    lines.push(current_line);
+                }
+
+                if should_break_word
+                    && get_text_width(word, &harf_font_ref, &skrifa_font_ref, font_size) > width
+                {
+                    let mut temp_word = String::new();
+                    for c in word.chars() {
+                        let next_temp_word = format!("{}{}", temp_word, c);
+                        if !temp_word.is_empty()
+                            && get_text_width(
+                                &next_temp_word,
+                                &harf_font_ref,
+                                &skrifa_font_ref,
+                                font_size,
+                            ) > width
+                        {
+                            lines.push(temp_word);
+                            temp_word = c.to_string();
+                        } else {
+                            temp_word = next_temp_word;
+                        }
+                    }
+                    current_line = temp_word;
+                } else {
+                    current_line = word.to_string();
+                }
+            }
+        }
+        if !current_line.is_empty() {
+            lines.push(current_line);
+        }
+        all_lines.extend(lines);
+    }
+
+    let total_height = all_lines.len() as f32 * line_height;
+
+    println!("Height: {}px\nLines: {:#?}", total_height, all_lines);
```

Comment:
probably don't want this in the committed copy?

Reply:
Removed